civicrm_entity.module

```diff
@@ -109,10 +115,108 @@ function civicrm_entity_entity_type_build(array &$entity_types) {
+      $entity_type_info['links']['add-form'] = sprintf('%s/{%s}', $entity_type_info['links']['add-form'], $entity_type_info['entity_keys']['bundle']);
+    }
+
     $entity_types[$entity_type_id] = new ContentEntityType($entity_type_info);
   }
 }
 
+/**
+ * Implements hook_entity_bundle_info().
+ */
+function civicrm_entity_entity_bundle_info() {
+  $transliteration = \Drupal::transliteration();
+  /** @var \Drupal\civicrm_entity\CiviCrmApiInterface $civicrm_api */
+  $civicrm_api = \Drupal::service('civicrm_entity.api');
+
+  $bundles = [];
+  $entity_types_with_bundles = array_filter(SupportedEntities::getInfo(), static function(array $civicrm_entity_info) {
+    return !empty($civicrm_entity_info['bundle property']);
+  });
+  foreach ($entity_types_with_bundles as $entity_type_id => $civicrm_entity_info) {
+    // We keep a bundle that is the same as the entity type ID. This allows us
+    // to create fields as if this entity has no bundles.
+    $bundles[$entity_type_id] = [
+      $entity_type_id => [
+        'label' => $civicrm_entity_info['civicrm entity label'],
+      ],
+    ];
+    $options = $civicrm_api->getOptions($civicrm_entity_info['civicrm entity name'], $civicrm_entity_info['bundle property']);
+    foreach ($options as $option) {
+      $machine_name = SupportedEntities::optionToMachineName($option, $transliteration);
+      $bundles[$entity_type_id][$machine_name]['label'] = $option;
+    }
+  }
+  return $bundles;
+}
+
+/**
+ * Implements hook_entity_bundle_field_info().
+ *
+ * This ensures CiviCRM Entity entity types have their field config instances
+ * across all bundles. It's a copy of the Field module's logic, but clones
+ * field config definitions.
+ *
+ * @see field_entity_bundle_field_info().
+ */
+function civicrm_entity_entity_bundle_field_info(EntityTypeInterface $entity_type, $bundle, array $base_field_definitions) {
+  $result = [];
+  if ($entity_type->get('civicrm_entity_ui_exposed') && $entity_type->hasKey('bundle')) {
+    // Query by filtering on the ID as this is more efficient than filtering
+    // on the entity_type property directly.
+    $ids = \Drupal::entityQuery('field_config')
+      ->condition('id', $entity_type->id() . '.', 'STARTS_WITH')
+      ->execute();
+    // Fetch all fields and key them by field name.
+    $field_configs = FieldConfig::loadMultiple($ids);
+
+    // Clone the field configs, so that we can modify them and change the
+    // target bundle type without manipulating the statically cached entries
+    // in the entity storage;
+    $cloned_field_configs = array_map(static function (FieldConfigInterface $field) use ($bundle) {
+      $cloned = clone $field;
+      $cloned->set('bundle', $bundle);
+      return $cloned;
+    }, $field_configs);
+    foreach ($cloned_field_configs as $field_instance) {
+      $result[$field_instance->getName()] = $field_instance;
+    }
+  }
+  return $result;
```

Comment:
We copy the logic in `field.module` to define bundle fields. The "trick" is that we load field configs for the entity type. We make clone of the object – so we don't pollute existing references – and change the bundle. This acts almost like a deriver and makes derivatives of each field config for each bundle.